core/common/src/main/java/alluxio/metrics/MetricsSystem.java

```diff
@@ -412,6 +415,53 @@ public static void resetAllCounters() {
     }
   }
 
+  /**
+   * This method is used to return a list of RPC metric objects which will be sent to the
+   * MetricsMaster.
+   *
+   * It is mainly useful for metrics which have a {@link MetricType} of COUNTER. The metric values
+   * with this type will be only include value of the difference since the last time that
+   * {@code reportMetrics} was called.
+   *
+   * By sending only the difference since the last RPC, this method will allow the master to
+   * track the metrics for a given worker, even if the worker is restarted.
+   */
+  private static List<alluxio.grpc.Metric> reportMetrics(InstanceType instanceType) {
+    List<alluxio.grpc.Metric> rpcMetrics = new ArrayList<>(20);
+    for (Metric m : allMetrics(instanceType)) {
+      // last reported metrics only need to be tracked for COUNTER metrics
+      // Store the last metric value which was sent, but the rpc metric returned should only
+      // contain the difference of the current value, and the last value sent. If it doesn't
+      // yet exist, just send the current value
+      if (m.getMetricType() == MetricType.COUNTER) {
+        Metric prev = LAST_REPORTED_METRICS.replace(m.getFullMetricName(), m);
+        // On restarts counters will be reset to 0, so whatever the value is the first time
+        // this method is called represents the value which should be added to the master's
+        // counter.
+        double diff = prev != null ? m.getValue() - prev.getValue() : m.getValue();
+        rpcMetrics.add(new Metric(m.getInstanceType(), m.getHostname(), m.getInstanceId(),
+            m.getMetricType(), m.getName(), diff).toProto());
+      } else {
+        rpcMetrics.add(m.toProto());
+      }
+    }
+    return rpcMetrics;
+  }
+
+  /**
+   * @return the worker metrics to send via RPC
+   */
+  public static List<alluxio.grpc.Metric> reportWorkerMetrics() {
+    return reportMetrics(InstanceType.WORKER);
+  }
+
+  /**
+   * @return the client metrics to send via RPC
+   */
+  public static List<alluxio.grpc.Metric> reportClientMetrics() {
+    return reportMetrics(InstanceType.CLIENT);
```

Comment:
How are the metrics currently tied to instance type? I think workers could potentially have client metrics due to using an Alluxio client for some operations under the hood?

Reply:
Yes, it is possible that workers also have "clients" but those clients should act like any other application and report their metrics via the client metrics heartbeat. However, that all hinges on the `allMetrics` method actually returning _only_ worker or client metrics.

Because it's static and will return everything for the JVM, we may need to do some more filtering....investigation required...

Reply:
Similarly, it also hinges on the fact that metrics registered within the JVM also are registered properly under Client, Master, Worker, etc.. 

When I'm working in my IDE this isn't implemented properly such that different Alluxio processes can live alongside one another in the metric system.

I think my vote is to actually just get rid of the static system entirely in this PR. It bothers me enough that I really just want to do away with it, especially if it is something on the radar anyways. Maybe an extra day of work to finish.